3di | Making and reviewing a pull request (#6892)
ros2/ros2_documentation: source/The-ROS2-Project/Contributing/Contributing-to-code/Reviewing-a-PR.rst

Reviewing a pull request (PR) — how-to
======================================

> All incoming code and documentation to ROS projects must be reviewed in a pull request.

Reviewer: We generally require two reviews. That might be worth mentioning up front.

Author: Thanks —I've put this right under the first bullet of section 1, as it seems to fit well there, rather than in this short description if that works for you.

Any developer is welcome to review and approve a pull request after review, when the changes are ready to merge.
Only a ROS core maintainer for the target repository can merge a pull request into that repository.

Prerequisites
-------------

A code or documentation contributor has made a pull request to merge their changes into one of `the ROS repositories <https://github.com/ros2>`__.

Steps
-----

1 Preparing for review
^^^^^^^^^^^^^^^^^^^^^^

* :ref:`Any developer <general-principles>` is welcome to review a pull request.
* Treat reviewing a pull request as a collaborative activity involving the submitter and other developers, rather than a passive or one-way process.
* As a reviewer:

  * You can make small improvements to code or documentation in-place, such as fixing typos or addressing minor style issues.
  * You should make a best effort attempt to comment on the pull request within one week of submission.

* When you begin `reviewing a pull request <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews>`__, leave a comment to let others know it is under review.

> 2 Reviewing the pull request

Reviewer: The problem with PR reviews isn't necessarily that people don't know how to do them, it is they don't know how to do them in a kind, respectful, and productive manner. I had a rough ROSCon talk outline that I put together and I was going to advocate that we move to a more concrete model for PR review comments.

There are roughly six types of comments that belong in a PR:  “Nit:”, "Issue:", "Suggestion:", "Question:", "Chore:","Praise/Thanks."

Reviewers should strive to provide clarity around: 
* Suggestions (i.e. things that might help but are not required for approval)
* High level concerns (i.e. refactorings). It helps to frame these as questions. (Questions and Issues)
* Low level concerns (line level changes ⇒ use Nit, Suggestion, or Chore)
* Reviewers should be clear on changes that block / do not block on merging. 

@christophebedard I'd be interested to hear your opinion here. Do you think we should be giving more guidance in those domain?

Reviewer 2: 100% agree; this section should cover this.

I would also mention trying to start with bigger concerns (e.g., wrong design) before starting to comment on minor concerns (e.g., variable naming and other nits). There's no point in nitpicking stuff if everything is going to change anyway because of some bigger design concern. This helps minimize the back and forth between the author and reviewer, which can be annoying and add to the overall time it takes for a PR to be reviewed.

Author: Thanks, folks -- this sounds very worthwhile. 

@kscottz It sounds like you are suggesting a formal system of a sort with the 6 types of comments, especially if your points on high level/low level concerns actually reference the comment types.

If you could provide or direct me to some input on that, I can make an attempt at adding some new stuff under section 2 along the lines discussed here.

Reviewer: I wouldn't quite think of these as a full fledged formal system, but rather a way to frame reviewing a pull request. They help clarify the expected outcome on the part of the reviewer. As you can probably tell we're not great at using them ourselves. 

At a high level, we want reviewers to be clear on issues that block and do not block getting a PR reviewed. 

* Nit (Nitpick) - tiny, non-blocking piece of feedback about style, naming, or readability and does block a PR from being merged, and the author can safely address or ignore it.
* Issue: highlights a specific problem like a bug or logical error, that should be paired with a suggested fix (possibly using the suggestion feature. Issues generally block getting a PR merged.
* Suggestion: an improvement to the current code structure, logic, or performance that aren't a style preference. These are nice to have but should not block getting a PR merged.
* Question: Asking for clarification to understand why a specific approach was taken. 
* Chore: Refers to work that doesn't add new features or fix bugs, but is required to keep the repository healthy. Roughly this is like asking someone to change your oil while they are also replacing your car's timing belt. It is stuff that needs to get done and it might as well get done when the hood of the car is already open.
* Praise: Self explanatory, but it helps offset the other stuff.

Author: Great, thanks for that. Will see what I can do and turn round another draft for review.

Author: @kscottz @christophebedard Hello - I've added some new material to these articles to cover this. 

We've simplified some of your input for the sake of brevity - all the information you've provided was great, but when placed all together it made the article seem overwhelming, particularly for new contributors. Hopefully this draft strikes the balance between being comprehensivem but also short and clear.

Reviewing a pull request (PR) — how-to
======================================

All incoming code and documentation to ROS projects must be reviewed in a pull request.
This article explains how to prepare for and review a pull request submitted by a contributor.
After reading this article, you'll be able to ensure changes in a pull request meet the required standards.

**Area: community | Content-type: how-to | Experience: beginner, intermediate, expert**

.. contents:: Table of Contents
   :depth: 2
   :local:

Summary
-------

Reviewing a pull request (PR) from a contributor allows you to check that their changes meet the appropriate guidelines and standards.

> Any developer is welcome to review and approve a pull request after review, when the changes are ready to merge.

Reviewer 2: This sentence feels clunky. It's saying two things:

1. Any developer is welcome to review and approve a pull request
2. Changes are ready to merge after a review/approval(?)

I think we can cut this sentence to only cover (1). And maybe make (2) a separate sentence if you really want to.

Only a ROS core maintainer for the target repository can merge a pull request into that repository.

Prerequisites
-------------

A code or documentation contributor has made a pull request to merge their changes into one of `the ROS repositories <https://github.com/ros2>`__.

Steps
-----

1 Preparing for review
^^^^^^^^^^^^^^^^^^^^^^

* :ref:`Any developer <general-principles>` is welcome to review a pull request.

  A pull request generally requires two reviews before it can be merged.

* Treat reviewing a pull request as a collaborative activity involving the submitter and other developers, rather than a passive or one-way process.
* As a reviewer:

  * You can make small improvements to code or documentation in-place, such as fixing typos or addressing minor style issues.

>   * You should make a best effort attempt to comment on the pull request within one week of submission.

Reviewer 2: > within one week of submission.

not sure we want to make a specific, official-sounding timeline commitment here other than say "within a reasonable amount of time."

@mjcarroll what do you think?

Reviewer 2: ah, I noticed the existing section did mention "one week." Then we can keep this for now and re-evaluate later.

* When you begin `reviewing a pull request <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/about-pull-request-reviews>`__, leave a comment to let others know it is under review.

2 Reviewing the pull request
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Use the following guidelines to review the submitted pull request:

* Confirm the code or documentation changes are appropriate for the repository.
* Verify the code is correct and complete, and scoped to a single, well defined change.
* Check that the pull request targets the default branch (usually ``rolling``).
* If the changes are based on a design document, verify that the changes are consistent with the design.
* For code changes, ensure that the changes:

    * Follow the :doc:`Developer guide <../Developer-Guide>`.
    * Follow the :doc:`Code style guide <../Code-Style-Language-Versions>`.
    * Include tests for the new feature or bug fix.

* For documentation changes, ensure the changes follow the :doc:`documentation guidance <../Contributing-To-ROS-2-Documentation>`.
* Confirm that the Continuous Integration (CI) run for the pull request passes cleanly.

> You can comment on the pull request for the submitter (`see the GitHub documentation for guidance <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request>`__), or suggest changes directly in the pull request.

Reviewer 2: That link covers suggesting changes using the UI button, so I would move the link to the end:

```suggestion
You can comment on the pull request for the submitter, or suggest changes directly in the pull request (`see the GitHub documentation for guidance <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request>`__).
```

3 Approving and merging the pull request
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

After you have reviewed the pull request and provided feedback, the submitter may continue the discussion or iterate on their changes, adding new commits to the PR.

When you are satisfied with the changes and they are ready to be merged, approve the pull request (`see the GitHub documentation for guidance <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews>`__).

> * :ref:`Any developer <general-principles>` is welcome to approve a pull request when the changes are ready to be merged.

Reviewer 2: Similar to my comment above, I don't get what this sentence is trying to say. I'd frame this as:

```suggestion
* :ref:`Any developer <general-principles>` is welcome to review a pull request, even if it already has a review.
```
otherwise you're just telling people to approve when someone else has already approved(?). I don't get it

Author: Thanks for this one -- in truth, I think I struggled to understand the flow of responsibilities here, so this is helpful.
